### ISSUE TITLE
System.Drawing.Bitmap.Clone: Why does Clone throws an OutOfMemoryException for Out-Of-Bounds problems

### ISSUE THREAD

**Issue by Timo-Weike** [dotnet/runtime#34395]
Why is it that `Bitmap.Clone` throws an `OutOfMemoryException` when the given `Rectangle` is not within the bounds of the source image. I don't see how this sort of argument error should be the cause of an `OutOfMemoryException`.

The method does check that the rectanlge has a non zero `Height` and `Width` and throws an `ArgumentException` if it does. But why is there not a similar check for the bounds of the rectanlge.

I could understand that the call to `Gdip.GdipCloneBitmapAreaI()` may could lead to an `OutOfMemoryException` since it is implemented in nativ code, but shouldn't this kind of error be handled as an `ArgumentException`.

At least the message of the `OutOfMemoryException` could be better, to pin point the actual problem.

Also according to the documentation of the `OutOfMemoryException`

> [The exception that is thrown when there is not enough memory to continue the execution of a program.](https://docs.microsoft.com/en-us/dotnet/api/system.outofmemoryexception?view=netcore-3.1)

this behavior is somewhat wrong as a console application on windows with this `Main`

``` cs
public static void Main()
{
	var bitmap = new Bitmap(100, 100);
	Rectangle rect = new Rectangle(90, 90, 20, 20);
	var part = bitmap.Clone(rect, bitmap.PixelFormat);
}
```

exits with an `OutOfMemoryException` but has only about 10MB of ram allocated to it.

So my proposal would be to add an additional argument check to the begining of the `Clone` methodes. Something like this

``` cs
[MethodImpl(MethodImplOptions.AggressiveInlining)]
static bool InRange(int x, int a, int b) => a <= x && a < b;

if (!InRange(rect.Left, 0, bitmap.Width)
	|| !InRange(rect.Right, 0, bitmap.Width)
	|| !InRange(rect.Top, 0, bitmap.Height)
	|| !InRange(rect.Bottom, 0, bitmap.Height)
	|| rect.Left >= rect.Right
	|| rect.Top >= rect.Height)
{
	throw new OutOfMemoryException($"The rectangle {rect} is outside of the source bitmap bounds.");
	// or better
	throw new ArgumentException($"The rectangle {rect} is outside of the source bitmap bounds.");
}
```

**Comment by gfoidl** [comment]
Debugging shows that `status` in https://github.com/dotnet/runtime/blob/676cb172347c8dfb0d214079add1b8ec6bf7f840/src/libraries/System.Drawing.Common/src/System/Drawing/Bitmap.cs#L314-L324 is `3` which maps to https://github.com/dotnet/runtime/blob/97af1e49abd7ec610b73c766f4066c173ec17173/src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs#L110-L111
https://github.com/dotnet/runtime/blob/97af1e49abd7ec610b73c766f4066c173ec17173/src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs#L75

The [mapping of the status codes](https://github.com/dotnet/runtime/blob/97af1e49abd7ec610b73c766f4066c173ec17173/src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs#L69-L91) matches the [Gdiplustypes.h](https://docs.microsoft.com/en-us/windows/win32/api/gdiplustypes/ne-gdiplustypes-status).

So per se this isn't an issue for .NET. Ideally this should be fixed by GDI+.

---
On non-windows with [libgdiplus](https://github.com/mono/libgdiplus) it's the [correct behavior](https://github.com/mono/libgdiplus/blob/2384cce9fef6f7bf9fd9c9aaca4d197eb1a31522/src/bitmap.c#L1121-L1122) as `InvalidParameter` is [defined with 2](https://github.com/mono/libgdiplus/blob/bbd132a0ad01b1f514836b40fe49e23e0ef6b3da/src/gdipenums.h#L344) and results in
```
 System.ArgumentException: Parameter is not valid.
```

**Comment by Timo-Weike** [comment]
Since this behavior is documented in the msdn docs for [`Bitmap.Clone`](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.bitmap.clone?view=netcore-3.1#System_Drawing_Bitmap_Clone_System_Drawing_Rectangle_System_Drawing_Imaging_PixelFormat_) for 3 years (according to [this](https://github.com/dotnet/dotnet-api-docs/blame/7f1ba7efad68821a2cd20a5cc8f64dee2ee35f27/xml/System.Drawing/Bitmap.xml#L680)). I don't know how high the chances are of this getting fixed in the Windows GDI+ version.

And so fixing atleast the bad exception message would be imo worth doing the change and also not break the documented behavior.

Or what would be the policy when the platform specific implementation of the same nativ interface differs like this?

**Comment by joshudson** [comment]
Random `OutOfMemoryException` from GDI+ has been the bane of .NET programmers since at least 2005. It's still a bug and causes no end of support departments chasing nonsense.

**Comment by danmosemsft** [comment]
Another option depending on your scenery is to use another drawing library like ImageSharp which is completely managed.

**Comment by JeremyKuhne** [comment]
Yes, we're unfortunately constrained by GDI+ here. The out of memory thing is a pet peeve of mine as well, but our hands are tied.
